src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py

```diff
@@ -0,0 +1,346 @@
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford, STFC Daresbury Lab
+
+'''Module providing a transformation from an Assignment node
+containing an Array Reference node in its left-hand-side which in turn
+has at least one PSyIR Range node specifying an access to an array
+index (equivalent to an array assignment statement in Fortran) to the
+equivalent loop representation using a NemoLoop node. A Range node is
+provided to the apply method of the tranformation to indicate which
+array index should be transformed.
+
+'''
+
+from __future__ import absolute_import
+
+from psyclone.psyGen import Transformation
+from psyclone.psyir.transformations.transformation_error \
+    import TransformationError
+from psyclone.errors import InternalError
+from psyclone.psyir.symbols import DataSymbol, INTEGER_TYPE
+from psyclone.psyir.nodes import Range, Reference, ArrayReference, \
+    Assignment, Literal, Operation, BinaryOperation
+from psyclone.nemo import NemoLoop, NemoKern
+from psyclone.configuration import Config
+
+
+class NemoArrayRange2LoopTrans(Transformation):
+    '''Provides a transformation from a PSyIR ArrayReference Range to a
+    PSyIR NemoLoop. For example:
+
+    >>> from psyclone.parse.algorithm import parse
+    >>> from psyclone.psyGen import PSyFactory
+    >>> api = "nemo"
+    >>> filename = "tra_adv.F90" # examples/nemo/code
+    >>> ast, invoke_info = parse(filename, api=api)
+    >>> psy = PSyFactory(api).create(invoke_info)
+    >>> schedule = psy.invokes.invoke_list[0].schedule
+    >>> schedule.view()
+    >>>
+    >>> from psyclone.psyir.nodes import Range
+    >>> from psyclone.domain.nemo.transformations import \
+    >>>     NemoArrayRange2LoopTrans
+    >>> from psyclone.transformations import TransformationError
+    >>>
+    >>> trans = NemoArrayRange2LoopTrans()
+    >>> for my_range in reversed(schedule.walk(Range)):
+    >>>     try:
+    >>>         trans.apply(my_range)
+    >>>     except TransformationError:
+    >>>         pass
+    >>> schedule.view()
+
+    '''
+    def apply(self, node, options=None):
+        '''Apply the NemoOuterArrayRange2Loop transformation if the supplied
+        node is the outermost Range node specifying an access to an
+        array index within an Array Reference that is on the
+        left-hand-side of an Assignment node. If this is the case then
+        the outermost Range nodes within array references within the
```

Comment:
Does this need re-writing now? A quick scan of the code doesn't seem to show any need for the supplied range to be outermost?

Reply:
The generated code would be incorrect if the transformation is applied to a Range node that is not outermost. This is checked in the validate() method.

Reply:
As you suggest later, I've tried to make this clearer in the documentation.